### PR TITLE
stm-dfu: fixes for flashing dfuse on STM32f745

### PR DESCRIPTION
**Pull request by amfern**

### CODE REVIEW

File: plugins/dfu/dfu-target-stm.c [hunk 1]
```diff
@@ -100,6 +100,9 @@ dfu_target_stm_upload_element (DfuTarget *target,
 	g_autoptr(GBytes) contents_truncated = NULL;
 	g_autoptr(GPtrArray) chunks = NULL;
 
+  /* this takes a long time on some devices */
```
Comment (hughsie) [inline plugins/dfu/dfu-target-stm.c:103]: tab here please

File: plugins/dfu/dfu-target-stm.c [hunk 2]
```diff
@@ -100,6 +100,9 @@ dfu_target_stm_upload_element (DfuTarget *target,
...
+	dfu_device_set_timeout (device, 5000);
+
```
Comment (hughsie) [inline plugins/dfu/dfu-target-stm.c:105]: Could we set this in dfu_target_stm_init() instead?

File: plugins/dfu/dfu-target-stm.c [hunk 3]
```diff
@@ -100,6 +100,9 @@ dfu_target_stm_upload_element (DfuTarget *target,
 	g_autoptr(GBytes) contents_truncated = NULL;
 	g_autoptr(GPtrArray) chunks = NULL;
 
+  /* this takes a long time on some devices */
+	dfu_device_set_timeout (device, 5000);
```
Comment (FFY00) [inline plugins/dfu/dfu-target-stm.c:104]: This should probably be in the metadata, no?

File: plugins/dfu/dfu-target-stm.c (hunk 2 above)
Comment (amfern) [reply plugins/dfu/dfu-target-stm.c:105]: yes, i can put it here

File: plugins/dfu/dfu-target-stm.c (hunk 3 above)
Comment (amfern) [reply plugins/dfu/dfu-target-stm.c:104]: ugh, what metadata? there is another way to configure this timeout?

Comment (hughsie) [reply plugins/dfu/dfu-target-stm.c:104]: At the moment we have `timeout_ms = 1500` which is a fairly long timeout for a USB transaction. @FFY00 is right that 5s is on the unusual side. Is the 5s timeout required for all DFU actions, or just one? For instance, does `GetStatus()` sometimes take a lot of time, or is it something more usual like `Detach()`? Follow up question -- is it all STM chips that need this long a timeout, or just the `STM32F745`? Is there an errata or specification change I've missed that specifies this long timout?

Thanks for any help!

Comment (amfern) [reply plugins/dfu/dfu-target-stm.c:104]: > s the 5s timeout required for all DFU actions, or just one? For instance, does GetStatus() sometimes take a lot of time, or is it something more usual like Detach()?


sorry i should have started with the problem, i tracked down the issue to `g_usb_device_control_transfer()` in `dfu_device_refresh()`, giving it a longer timeout solved the issue
```
device-erase:	0%
device-write:	0%
device-write:	100%
idle:	100%
device-write:	100%
device-write:	0%
device-write:	100%
idle:	100%
device-verify:	100%
device-write:	100%
idle:	100%
device-erase:	100%
device-write:	100%
device-write:	20%
device-write:	40%
device-write:	60%
device-write:	80%
cannot erase address 0x8040000: cannot get device state: transfer timed out
```

> is it all STM chips that need this long a timeout
I don't know, i have only STM32F745 and STM32F745VGH6, and i observed the same behavior with both

> Is there an errata or specification change I've missed that specifies this long timout?
sorry i don't know, its just the behavior i observed

Comment (amfern) [reply plugins/dfu/dfu-target-stm.c:104]: full log
```(dfu-tool:3465806): FuCommon-DEBUG: 11:20:22.685: reading obj/xtend_fcos_2.2.0-RC3_KAKUTEF7-2.dfu with 366830 bytes
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.704: STM-DFU support
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.704: using DFU transfer size 0x0040 bytes
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.704: STM-DFU support
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.704: using DFU transfer size 0x0040 bytes
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.704: STM-DFU support
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.704: using DFU transfer size 0x0040 bytes
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.704: STM-DFU support
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.704: using DFU transfer size 0x0800 bytes
(dfu-tool:3465806): FuDevice-DEBUG: 11:20:22.704: using 301046452a49d84af6356d23e43a684b8f10660f for usb:01:00:02
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.706: Zone:0, Sec#:0, Addr:0x08000000, Size:0x8000, Caps:0x7 [REW]
Zone:0, Sec#:0, Addr:0x08008000, Size:0x8000, Caps:0x7 [REW]
Zone:0, Sec#:0, Addr:0x08010000, Size:0x8000, Caps:0x7 [REW]
Zone:0, Sec#:0, Addr:0x08018000, Size:0x8000, Caps:0x7 [REW]
Zone:0, Sec#:1, Addr:0x08020000, Size:0x20000, Caps:0x7 [REW]
Zone:0, Sec#:2, Addr:0x08040000, Size:0x40000, Caps:0x7 [REW]
Zone:0, Sec#:2, Addr:0x08080000, Size:0x40000, Caps:0x7 [REW]
Zone:0, Sec#:2, Addr:0x080c0000, Size:0x40000, Caps:0x7 [REW]
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.707: Zone:0, Sec#:0, Addr:0x1fff0000, Size:0x0020, Caps:0x3 [RW]
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.707: Zone:0, Sec#:0, Addr:0x1ff0f000, Size:0x0400, Caps:0x3 [RW]
Zone:0, Sec#:1, Addr:0x1ff0f400, Size:0x0020, Caps:0x3 [RW]
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.708: Zone:0, Sec#:0, Addr:0xffff0000, Size:0x0004, Caps:0x3 [RW]
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.708: refreshed status=errFIRMWARE and state=dfuERROR (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.708: refreshed status=errFIRMWARE and state=dfuERROR (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.708: clearing error errFIRMWARE
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.711: downloading to target: @Internal Flash  /0x08000000/04*032Kg,01*128Kg,03*256Kg
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.711: downloading element at 0x8000000
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.711: marking sector 0x8000000-8008000 to be erased
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.711: setting action device-erase
device-erase:	0%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.711: erasing sector at 0x8000000
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.711: refreshed status=OK and state=dfuDNBUSY (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:22.711: sleeping for 5ms…
device-write:	0%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.067: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.067: doing actual check status
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.067: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.067: setting percentage 100% of device-erase
device-write:	100%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.067: setting action idle
idle:	100%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.067: setting action device-write
device-write:	100%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.067: setting address to 0x8000000
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.068: refreshed status=OK and state=dfuDNBUSY (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.068: sleeping for 5ms…
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.073: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.073: doing actual check status
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.074: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.074: writing sector at 0x8000000 (0x728)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.076: refreshed status=OK and state=dfuDNBUSY (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.076: sleeping for 5ms…
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.088: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.088: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.088: setting percentage 0% of device-write
device-write:	0%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.088: setting percentage 100% of device-write
device-write:	100%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.088: setting action idle
idle:	100%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.089: setting action device-verify
device-verify:	100%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.089: using sector 0 for read of 8000000
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.089: ignoring action device-read as device-verify already set and not idle
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.089: setting DfuSe address to 0x8000000
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.089: refreshed status=OK and state=dfuDNBUSY (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.089: sleeping for 5ms…
device-write:	100%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.094: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.094: doing actual check status
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.095: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.098: got #0000 chunk @0x8000000 of size 2048
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.098: setting action idle
idle:	100%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.098: downloading element at 0x8010000
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.098: marking sector 0x8008000-8010000 to be erased
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.098: marking sector 0x8010000-8018000 to be erased
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.098: marking sector 0x8018000-8020000 to be erased
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.098: marking sector 0x8020000-8040000 to be erased
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.098: marking sector 0x8040000-8080000 to be erased
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.098: setting action device-erase
device-erase:	100%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.098: erasing sector at 0x8008000
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.099: refreshed status=OK and state=dfuDNBUSY (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.099: sleeping for 5ms…
device-write:	100%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.443: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.443: doing actual check status
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.444: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.444: setting percentage 20% of device-erase
device-write:	20%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.444: erasing sector at 0x8010000
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.445: refreshed status=OK and state=dfuDNBUSY (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.445: sleeping for 5ms…
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.814: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.814: doing actual check status
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.815: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.815: setting percentage 40% of device-erase
device-write:	40%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.815: erasing sector at 0x8018000
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.815: refreshed status=OK and state=dfuDNBUSY (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:23.815: sleeping for 5ms…
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:24.156: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:24.156: doing actual check status
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:24.156: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:24.156: setting percentage 60% of device-erase
device-write:	60%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:24.156: erasing sector at 0x8020000
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:24.157: refreshed status=OK and state=dfuDNBUSY (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:24.157: sleeping for 5ms…
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:25.225: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:25.225: doing actual check status
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:25.225: refreshed status=OK and state=dfuDNLOAD-IDLE (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:25.225: setting percentage 80% of device-erase
device-write:	80%
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:25.225: erasing sector at 0x8040000
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:25.226: refreshed status=OK and state=dfuDNBUSY (dnload=5)
(dfu-tool:3465806): FuPluginDfu-DEBUG: 11:20:25.226: sleeping for 5ms…
cannot erase address 0x8040000: cannot get device state: transfer timed out```

**Comment by amfern** [comment]
now this branch is using the DfuForceTimeout feature from #1946

**Pull request merged**